[AIRFLOW-2694] Declare permissions in DAG definition

astahlman opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [ X ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2694
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [ X ] Here are some details about my PR, including screenshots of any UI changes:

This PR adds support for declaratively assigning DAG-level permissions
to a role via the `DAG.__init__()` method.

When the DAG definition is evaluated and the `access_control` argument
is supplied, we update the permissions on the ViewMenu associated with
this DAG according to the following rules:

- If the role does not exist, we raise an exception.
- If the role exists, we ensure that it has the specified set of
  permissions on the DAG
- If any other permissions exist for the DAG that are not specified in
  `access_control`, we revoke them

### Tests

- [ X ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added unit tests in `tests/models.py` to test that:

- Invalid roles raise an exception
- Invalid permissions raise an exception
- New permissions in the access_control map are created
- Existing permissions that are not specified in the access_control map are revoked

### Commits

- [ X ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ X ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

Not sure if we have any docs that need updating besides the docstrings on the DAG class. If so, let me know and I'll be happy to update those, as well.

### Code Quality

- [ X ] Passes `flake8`

Commit:
[AIRFLOW-2694] Declare permissions in DAG definition

This PR adds support for declaratively assigning DAG-level permissions
to a role via the `DAG.__init__()` method.

When the DAG definition is evaluated and the `access_control` argument
is supplied, we update the permissions on the ViewMenu associated with
this DAG according to the following rules:

- If the role does not exist, we raise an exception.
- If the role exists, we ensure that it has the specified set of
  permissions on the DAG
- If any other permissions exist for the DAG that are not specified in
  `access_control`, we revoke them

Commit:
Move RBAC constants to break circular dependency

Commit:
Add license header

feng-tao commented on airflow/models/__init__.py:
```diff
@@ -4244,6 +4254,63 @@ def _test_cycle_helper(self, visit_map, task_id):
+        :type access_control: dict
+
+        """
+        sm = cached_appbuilder().sm
```
this may be a little concern as it requires starting the webserver when Airflow starts the scheduler.

Two ideas could help solve these cases:
1. move all FAB models(https://github.com/dpgaspar/Flask-AppBuilder/blob/1e900bba85452de6d988f7da191f9a26fec62226/flask_appbuilder/security/sqla/models.py) into Airflow code base(cc @mistercrunch  who brought up this idea previously). This helps Airflow not relied on FAB that much. 
2. Created an API endpoint in web server for creating permission / delete permission purpose(https://issues.apache.org/jira/browse/AIRFLOW-2693). In this way, the Airflow web server handles all the FAB permission actions. And scheduler just delegates the permission handling to the webserver.

cc @jgao54  @mistercrunch

astahlman replied:
@feng-tao Yeah, I had the same thought - I really don't like doing this in the `__init__` method.

I see that we are already doing some permission handling in the `/refresh` endpoint - maybe this would be a natural place to invoke `dag.set_dag_permissions()`?

https://github.com/apache/airflow/blob/81f7c475a92b40320a683fc0bf81b62c47052238/airflow/www/views.py#L1682

We could have similar logic in the `sync_perm` CLI command: for each DAG in the DagBag, call `dag.set_dag_permissions()`

https://github.com/apache/airflow/blob/81f7c475a92b40320a683fc0bf81b62c47052238/airflow/bin/cli.py#L1634

feng-tao replied:
I think extending the `/refresh` endpoint should work. We need to pass the `access_control` info from scheduler to webserver endpoint during the DAG `_init_` constructor.  I am not so sure how to put the `set_dag_permissions` function inside `sync_perms` cli though.

I think we couldn't get the dag access_control information until scheduler parses the dag information. And we need to persist these infomation somewhere in the db. Then the cli knows how to persist the permission with roles mapping information, right? In this case, I don't know how the CLI could figure how to call the ``sync_perm_for_dag``.

feng-tao replied:
sync offline, and @astahlman will move the logic into the `sync_perm` cli. This is to avoid putting extra latency into scheduler loop.

astahlman replied:
Updated - I've moved this logic from the DAG class to the AirflowSecurityManager and triggered it from the CLI `sync_perm` command as well as the `/refresh` and `/refresh_all` endpoints. This way is much better I think - thanks for the feedback.

Commit:
Sync DAG permissions via CLI and /refresh* endpoints

Move the DAG-level permission syncing logic into
`AirflowSecurityManager.sync_perm_for_dag`, and trigger this method from
the CLI's `sync_perm` command and from the `/refresh` and `/refresh_all`
web endpoints.

feng-tao reviewed (approved):
LGTM, thanks. Let's wait for a few days in case @jgao54 and @mistercrunch have further comment.

feng-tao commented on airflow/www/security.py:
```diff
@@ -451,19 +452,90 @@ def sync_roles(self):
         self.clean_perms()
 
-    def sync_perm_for_dag(self, dag_id):
+    def sync_perm_for_dag(self, dag_id, access_control):
```
How about ``def sync_perm_for_dag(self, dag_id, access_control=None):`` ?

astahlman replied:
Sure, updated.

Commit:
Default access_control to None
